### PR TITLE
Add support for preflight check configuration

### LINKED ISSUE #899: Add configuration support to kapp for preflight checks
**Describe the problem/challenge you have**
[A description of the current challenge that you are experiencing.]

Preflight checks have no method of configuration. This adds a generic configuration for preflight checks.

**Describe the solution you'd like**
[A clear and concise description of what you want to happen. If applicable a visual representation of the UX (ex: new CLI argument name, the behavior expected).]

This should implement https://github.com/carvel-dev/carvel/pull/729

**Anything else you would like to add:**
[Additional information that will assist in solving the issue.]

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by tmshort**
#### What this PR does / why we need it:

This adds configuration file support (`preflightRules`) for preflight checks. Each preflight check must add a `SetConfig()` method to accept configuration from the Preflight registry.

#### Which issue(s) this PR fixes:

Fixes #899

#### Does this PR introduce a user-facing change?

```release-note
Add support for preflightRules configuration
```

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs
```
Proposal https://github.com/carvel-dev/carvel/pull/729

### CODE REVIEW

File: pkg/kapp/preflight/registry.go
Comment (everettraven) [inline pkg/kapp/preflight/registry.go]: @tmshort I know we have already had this conversation in the PR against my fork before the CRD proposal was approved, but I wanted to call it out here and get the opinions of the carvel maintainers as well.

IMO we should change this to not accept the full `config.Conf` type. Instead I think something like:
```suggestion
func (c *Registry) SetRules(rules ...config.PreflightRule) error {
```
would be better here for two reasons:
1. This method really only cares about the preflight rules anyways. Because of this it is easier and more clear to a user what information is really used and taken into account when using this method to configure a preflight registry. Keeping the parameters to the minimum thing needed is a better UX IMO.
2. Eliminates the parsing of YAML in the unit tests, reducing the, IMO, unrelated points of failure in the unit tests. I feel that changes to logic that would break the YAML parsing should break tests specifically associated with parsing of the YAML contents and not the tests for a function that assumes the parsing has already been done and takes in the type resulting from the YAML parsing.

I'm by no means saying this can't go in like this, I'm totally fine with that if the consensus is that what is currently implemented is the best approach. Just wanted to highlight what I think might make things a bit easier to understand and reason about as both a consumer of the library and a maintainer.

Comment (tmshort) [reply pkg/kapp/preflight/registry.go]: _If_ this were the desired route, I'd prefer `[]PrelightRule` as the parameter, as that's the output of `config.PreflightRules()`. The above approach requires an unnecessary break-apart step, vs:
```
registry.SetConfig(conf.PreflightRules())
...
func (c *Registry) SetConfig(rules []PreflightRules) error {
...
```
I prefer the YAML config in the unit tests, as I think it's easier to read than golang structs.

Comment (everettraven) [reply pkg/kapp/preflight/registry.go]: > I'd prefer []PrelightRule as the parameter, as that's the output of config.PreflightRules(). The above approach requires an unnecessary break-apart step, vs

Fair point, totally fine with me. I was just throwing something together as an example.

>I prefer the YAML config in the unit tests, as I think it's easier to read than golang structs.

Also fair, but I personally feel its easier to read the golang structs than the yaml in this case. To each their own :). This also isn't explicitly about removing the YAML config, I'm fine with that approach if the consensus is that the currently implemented approach. I personally think the removal of the YAML config is one benefit, but I think the other benefit I mentioned stands on its own.

Comment (tmshort) [reply pkg/kapp/preflight/registry.go]: Yes, one could be done with (or without) the other.

Comment (praveenrewar) [reply pkg/kapp/preflight/registry.go]: Personally, I prefer the YAML config in the tests as well, as it's easier to scan through them, however, I agree with @everettraven that we need not pass the whole config and just the preflight rules should be enough.

**Review by praveenrewar — COMMENTED** [review]
Looks good to me apart from Bryce's comment. 
Should we add/update the preflight permissions e2e tests with a config example? (I am okay with just the unit tests as well)

Would would also have to add the preflight option in the [configuration](https://carvel.dev/kapp/docs/develop/config/) docs.

File: pkg/kapp/preflight/registry.go
Comment (tmshort) [reply pkg/kapp/preflight/registry.go]: I can tweak the API to only pass the preflight rules.

**Comment by tmshort** [comment]
> Looks good to me apart from Bryce's comment. Should we add/update the preflight permissions e2e tests with a config example? (I am okay with just the unit tests as well)
> 
> Would would also have to add the preflight option in the [configuration](https://carvel.dev/kapp/docs/develop/config/) docs.

I'll create a new PR for the docs.

The unit tests cover the code as written. IMHO, an e2e would be more appropriate once a preflight check adds configuration settings; otherwise, the e2e would effectively be a duplicate test.

**Comment by tmshort** [comment]
Documentation PR: https://github.com/carvel-dev/carvel/pull/740

**Review by everettraven — APPROVED** [review]

**Review by praveenrewar — APPROVED** [review]
LGTM! Thank you so much for the contribution ❤️

**Pull request merged**
